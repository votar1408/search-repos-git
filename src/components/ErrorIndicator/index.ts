import {ErrorIndicator} from './ErrorIndicator';

export {ErrorIndicator};
