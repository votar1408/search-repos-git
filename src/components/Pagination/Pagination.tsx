import './pagination.scss';

import React, {
    FC,
    ReactElement,
    useCallback,
    useMemo
} from 'react';

import {
    useDispatch,
    useSelector
} from 'react-redux';

import {common} from '../../consts';
import {RepositoryServerInfo} from '../../interfaces/ServerData';
import {changePageAction} from '../../redux/app/actions';
import {selectApp} from '../../redux/types';

export const Pagination: FC = () => {
    const {page, repositories, searchValue, loading} = useSelector(selectApp);
    const dispatch = useDispatch();

    const listForPage: RepositoryServerInfo[] = !repositories[page] ? repositories[page - 1] : repositories[page];

    const onChangePageHandler = useCallback(
        (value: number) => {
            dispatch(changePageAction(value));
        },
        [dispatch]
    );

    const buttons: ReactElement | null = useMemo(() => {
        return searchValue && repositories.length > 0 ? (
            <div className="pagination">
                <button
                    type="submit"
                    className="pagination__button"
                    disabled={page === common.getDefaultPage || loading}
                    data-id="left-button"
                    onClick={() => onChangePageHandler(-1)}
                >
                    <i className="fa fa-angle-left"></i>
                </button>
                <button
                    type="submit"
                    className="pagination__button"
                    disabled={!listForPage?.length || loading}
                    data-id="right-button"
                    onClick={() => onChangePageHandler(1)}
                >
                    <i className="fa fa-angle-right"></i>
                </button>
            </div>
        ) : null;
    }, [listForPage, loading, onChangePageHandler, page, repositories.length, searchValue]);

    return <>{buttons}</>;
};
