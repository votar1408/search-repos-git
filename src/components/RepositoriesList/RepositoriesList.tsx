import './repositoriesList.scss';

import React, {
    FC,
    ReactElement,
    useMemo
} from 'react';

import {useSelector} from 'react-redux';

import {texts} from '../../consts/texts';
import {RepositoryServerInfo} from '../../interfaces/ServerData';
import {selectApp} from '../../redux/types';
import {Repository} from './Repository';

export const RepositoriesList: FC = () => {
    const {repositories, page, loading} = useSelector(selectApp);
    const listForPage: RepositoryServerInfo[] = !repositories[page] ? repositories[page - 1] : repositories[page];

    const listItems: ReactElement | null = useMemo(() => {
        return listForPage?.length > 0 ? (
            <ul className="repositories-container">
                {listForPage.map((item: RepositoryServerInfo) => {
                    return <Repository key={item.id} item={item} />;
                })}
            </ul>
        ) : loading ? null : (
            <div className="empty-repositories">{texts.emptyList}</div>
        );
    }, [listForPage, loading]);

    return <>{listItems}</>;
};
