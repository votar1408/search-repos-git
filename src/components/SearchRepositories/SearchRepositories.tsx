import './searchRepositories.scss';

import React, {
    ChangeEvent,
    FC,
    FormEvent,
    useCallback,
    useEffect,
    useState
} from 'react';

import {
    useDispatch,
    useSelector
} from 'react-redux';

import {
    common,
    texts
} from '../../consts';
import {
    changeSearchValueAction,
    clearRepositoriesAction,
    getRepositoriesAction
} from '../../redux/app/actions';
import {selectApp} from '../../redux/types';
import {getRepositoriesUrl} from '../../routes';
import {
    useHttp,
    useMountedRef
} from '../hooks';

export const SearchRepositories: FC = () => {
    const dispatch = useDispatch();
    const isMountedRef = useMountedRef();
    const {page, repositories, searchValue} = useSelector(selectApp);
    const {request, controller} = useHttp();
    const [searchValueState, setSearchValueState] = useState<string>('');

    useEffect(() => {
        onSendRequestHandler();
        return () => controller.abort();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [page]);

    const onSendRequestHandler = useCallback(async () => {
        const isNewSearchValue: boolean = searchValueState !== searchValue;
        if (searchValueState && (isNewSearchValue || !repositories[page])) {
            try {
                const data = await request(
                    getRepositoriesUrl(searchValueState, isNewSearchValue ? common.getDefaultPage : page),
                    'GET'
                );

                if (isMountedRef.current) {
                    dispatch(getRepositoriesAction(data.items));
                }
            } catch (e) {
                alert(e.message);
            }
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [dispatch, searchValue, searchValueState, page, repositories]);

    const onChangeSearchValueHandler = useCallback((event: ChangeEvent<HTMLInputElement>) => {
        event.persist();
        setSearchValueState(event.target.value);
    }, []);

    const onSearchRepositoriesHandler = useCallback(
        async (event: FormEvent<HTMLFormElement>) => {
            event.preventDefault();

            onSendRequestHandler();
            if (isMountedRef.current) {
                if (searchValueState !== searchValue) {
                    dispatch(changeSearchValueAction(searchValueState));
                    dispatch(clearRepositoriesAction());
                }
            }
        },
        [dispatch, isMountedRef, onSendRequestHandler, searchValue, searchValueState]
    );

    return (
        <form className="search" onSubmit={onSearchRepositoriesHandler}>
            <input
                type="text"
                name="search-repo-input"
                className="search__term"
                placeholder={texts.inputRepositoryName}
                value={searchValueState}
                onChange={onChangeSearchValueHandler}
            />
            <button type="submit" name="submit-button" className="search__button">
                <i className="fa fa-search"></i>
            </button>
        </form>
    );
};
