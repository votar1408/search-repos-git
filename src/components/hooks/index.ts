import {useHttp} from './useHttp';
import {useMountedRef} from './useMountedRef';

export {useHttp, useMountedRef};
