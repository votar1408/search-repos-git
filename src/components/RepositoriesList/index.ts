import {RepositoriesList} from './RepositoriesList';

export {RepositoriesList};
