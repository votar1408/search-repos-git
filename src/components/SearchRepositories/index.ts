import {SearchRepositories} from './SearchRepositories';

export {SearchRepositories};
