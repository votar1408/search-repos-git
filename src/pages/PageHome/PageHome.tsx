import React, {FC} from 'react';

import {Pagination} from '../../components/Pagination/Pagination';
import {RepositoriesList} from '../../components/RepositoriesList/RepositoriesList';
import {SearchRepositories} from '../../components/SearchRepositories/SearchRepositories';

export const PageHome: FC = () => {
    return (
        <>
            <SearchRepositories />
            <RepositoriesList />
            <Pagination />
        </>
    );
};
