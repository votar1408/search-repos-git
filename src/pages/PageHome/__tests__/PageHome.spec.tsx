import React from 'react';

import {
    shallow,
    ShallowWrapper
} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';

import {Pagination} from '../../../components/Pagination';
import {RepositoriesList} from '../../../components/RepositoriesList';
import {SearchRepositories} from '../../../components/SearchRepositories';
import {PageHome} from '../PageHome';

describe('PageHome', () => {
    it('should render correctly', () => {
        const output = shallow(<PageHome />);
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        expect(shallow(<PageHome />));
    });

    it('should contains of <SearchRepositories/>', () => {
        const wrapper: ShallowWrapper = shallow(<PageHome />);
        expect(wrapper.find(SearchRepositories)).toHaveLength(1);
    });

    it('should contains of <RepositoriesList/>', () => {
        const wrapper: ShallowWrapper = shallow(<PageHome />);
        expect(wrapper.find(RepositoriesList)).toHaveLength(1);
    });

    it('should contains of <Pagination/>', () => {
        const wrapper: ShallowWrapper = shallow(<PageHome />);
        expect(wrapper.find(Pagination)).toHaveLength(1);
    });
});
