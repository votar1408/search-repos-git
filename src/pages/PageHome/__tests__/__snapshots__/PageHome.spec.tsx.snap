// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHome should render correctly 1`] = `
<Fragment>
  <SearchRepositories />
  <RepositoriesList />
  <Pagination />
</Fragment>
`;
